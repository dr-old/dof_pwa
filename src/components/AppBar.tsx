import { hexToRgba } from "../utils/helpers";
import {
  MenuOpenRounded,
  NotificationsOutlined,
  SearchRounded,
} from "@mui/icons-material";
import { Box, IconButton, Toolbar, Typography } from "@mui/material";
import { AppBarProps as MuiAppBarProps } from "@mui/material/AppBar";
import TextField from "@mui/material/TextField";
import { styled } from "@mui/material/styles";
import { useLocation } from "react-router-dom";

interface HeaderProps extends MuiAppBarProps {
  open?: boolean;
}

const drawerWidth = 240;

// MuiAppBar

const Header = styled(Box, {
  shouldForwardProp: (prop) => prop !== "open",
})<HeaderProps>(({ theme, open }) => ({
  transition: theme.transitions.create(["margin", "width"], {
    easing: theme.transitions.easing.sharp,
    duration: theme.transitions.duration.leavingScreen,
  }),
  ...(open && {
    width: `calc(100% - ${drawerWidth}px)`,
    transition: theme.transitions.create(["margin", "width"], {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
    marginRight: drawerWidth,
  }),
  boxShadow: "none", // Remove shadow
  backgroundColor: theme.palette.background.default, // Ensure background color is set
  backgroundImage: "none", // Ensure no background image is applied
}));

const OutlinedIconButton = styled(IconButton)(({ theme }) => ({
  border: `1px solid ${hexToRgba(theme.palette.text.secondary, 0.6)}`,
  borderRadius: theme.shape.borderRadius * 2.5,
  padding: theme.spacing(1),
  color: theme.palette.text.secondary,
  opacity: 0.6,
  "&:hover": {
    backgroundColor: theme.palette.action.hover,
  },
}));

const BorderlessTextField = styled(TextField)(({ theme }) => ({
  "& .MuiOutlinedInput-root": {
    "& fieldset": {
      border: "none", // Remove the border
    },
    "&:hover fieldset": {
      border: "none", // Ensure no border on hover
    },
    "&.Mui-focused fieldset": {
      border: "none", // Ensure no border when focused
    },
    "& input": {
      fontSize: "0.75rem", // Set font size for input text
    },
  },
  "& .MuiInputLabel-root": {
    fontSize: "0.75rem", // Set font size for the label if needed
  },
}));

interface AppBarProps {
  isMenuOpen?: boolean;
  toggleMenu?: () => void;
  isMobile?: boolean;
}

const AppBar = ({ isMenuOpen, toggleMenu, isMobile }: AppBarProps) => {
  const { state } = useLocation();

  return (
    <Header sx={{ backgroundColor: "transparent" }}>
      <Toolbar>
        <Typography variant="h6" noWrap sx={{ flexGrow: 1 }}>
          {state.title}
        </Typography>
        <BorderlessTextField
          variant="outlined"
          size="small"
          placeholder="Search something..."
          InputProps={{
            startAdornment: (
              <IconButton>
                <SearchRounded fontSize="small" />
              </IconButton>
            ),
          }}
          sx={{ mx: 2 }}
        />
        <OutlinedIconButton color="inherit">
          <NotificationsOutlined fontSize="small" />
        </OutlinedIconButton>
        {isMobile && (
          <OutlinedIconButton
            color="inherit"
            aria-label="open drawer"
            edge="start"
            onClick={toggleMenu}
            sx={{ marginLeft: 1 }}>
            <MenuOpenRounded fontSize="small" />
          </OutlinedIconButton>
        )}
      </Toolbar>
    </Header>
  );
};

export default AppBar;
