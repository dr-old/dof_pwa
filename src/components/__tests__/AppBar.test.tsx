// src/components/__tests__/AppBar.test.tsx
import { render, screen, fireEvent } from "@testing-library/react";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import AppBar from "../AppBar";
import { BrowserRouter as Router, useLocation } from "react-router-dom";
import "@testing-library/jest-dom";

// Create a custom theme to use in tests
const theme = createTheme({
  palette: {
    mode: "light",
    background: { default: "#ffffff" },
    text: { secondary: "#000" },
    action: { hover: "#e0e0e0" },
  },
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: jest.fn(),
}));

// Mock functions
const mockToggleMenu = jest.fn();

const renderComponent = (props = {}) => {
  return render(
    <ThemeProvider theme={theme}>
      <Router>
        <AppBar {...props} />
      </Router>
    </ThemeProvider>
  );
};

describe("AppBar Component", () => {
  test("renders correctly with menu and search", () => {
    (useLocation as jest.Mock).mockReturnValue({
      state: { title: "Dashboard" },
    });

    const { asFragment } = renderComponent({
      isMenuOpen: false,
      toggleMenu: mockToggleMenu,
      isMobile: true,
    });

    // Snapshot Test
    expect(asFragment()).toMatchSnapshot();

    // Check if title is rendered - adjust if state.title is set
    expect(screen.getByRole("heading", { level: 6 })).toBeInTheDocument();

    // Check if search icon button is present
    expect(screen.getByRole("button", { name: /search/i })).toBeInTheDocument();

    // Click on the search icon to open search
    fireEvent.click(screen.getByRole("button", { name: /search/i }));

    // Check if the search input field is rendered
    expect(
      screen.getByPlaceholderText("Search something...")
    ).toBeInTheDocument();

    // Click on the close icon to close search
    fireEvent.click(screen.getByRole("button", { name: /close/i }));

    // Check if the search input field is no longer rendered
    expect(
      screen.queryByPlaceholderText("Search something...")
    ).not.toBeInTheDocument();
  });

  test("calls toggleMenu function when menu button is clicked", () => {
    (useLocation as jest.Mock).mockReturnValue({
      state: { title: "Dashboard" },
    });

    renderComponent({
      isMenuOpen: false,
      toggleMenu: mockToggleMenu,
      isMobile: true,
    });

    // Click the menu button
    fireEvent.click(screen.getByRole("button", { name: /open drawer/i }));

    // Check if toggleMenu was called
    expect(mockToggleMenu).toHaveBeenCalled();
  });

  test("renders correctly without search and menu", () => {
    (useLocation as jest.Mock).mockReturnValue({
      state: { title: "Dashboard" },
    });

    const { asFragment } = renderComponent({
      isMenuOpen: false,
      toggleMenu: mockToggleMenu,
      isMobile: false,
    });

    // Snapshot Test
    expect(asFragment()).toMatchSnapshot();

    // Check if the search and menu buttons are not rendered
    expect(
      screen.queryByRole("button", { name: /search/i })
    ).not.toBeInTheDocument();
    expect(
      screen.queryByRole("button", { name: /open drawer/i })
    ).not.toBeInTheDocument();
  });
});
