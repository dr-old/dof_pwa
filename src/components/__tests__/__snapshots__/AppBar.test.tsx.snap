// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar Component renders correctly with menu and search 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-jes66t"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-1dvstpp-MuiTypography-root"
      >
        Dashboard
      </h6>
      <button
        aria-label="open search"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-ih405l-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="SearchRoundedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27c1.2-1.4 1.82-3.31 1.48-5.34-.47-2.78-2.79-5-5.59-5.34-4.23-.52-7.79 3.04-7.27 7.27.34 2.8 2.56 5.12 5.34 5.59 2.03.34 3.94-.28 5.34-1.48l.27.28v.79l4.25 4.25c.41.41 1.08.41 1.49 0 .41-.41.41-1.08 0-1.49zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-1mb0ci6-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="NotificationsOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.9 2 2 2m6-6v-5c0-3.07-1.63-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.64 5.36 6 7.92 6 11v5l-2 2v1h16v-1zm-2 1H8v-6c0-2.48 1.51-4.5 4-4.5s4 2.02 4 4.5z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="open drawer"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1p1z4ns-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="MenuOpenRoundedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 18h11c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1s.45 1 1 1m0-5h8c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1s.45 1 1 1M3 7c0 .55.45 1 1 1h11c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1m17.3 7.88L17.42 12l2.88-2.88c.39-.39.39-1.02 0-1.41a.9959.9959 0 0 0-1.41 0L15.3 11.3c-.39.39-.39 1.02 0 1.41l3.59 3.59c.39.39 1.02.39 1.41 0 .38-.39.39-1.03 0-1.42"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppBar Component renders correctly without search and menu 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-jes66t"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-1dvstpp-MuiTypography-root"
      >
        Dashboard
      </h6>
      <div
        class="MuiFormControl-root MuiTextField-root css-545qvj-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="SearchRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27c1.2-1.4 1.82-3.31 1.48-5.34-.47-2.78-2.79-5-5.59-5.34-4.23-.52-7.79 3.04-7.27 7.27.34 2.8 2.56 5.12 5.34 5.59 2.03.34 3.94-.28 5.34-1.48l.27.28v.79l4.25 4.25c.41.41 1.08.41 1.49 0 .41-.41.41-1.08 0-1.49zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart css-1ua80n0-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            placeholder="Search something..."
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-1mb0ci6-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="NotificationsOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.9 2 2 2m6-6v-5c0-3.07-1.63-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.64 5.36 6 7.92 6 11v5l-2 2v1h16v-1zm-2 1H8v-6c0-2.48 1.51-4.5 4-4.5s4 2.02 4 4.5z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
